### PR TITLE
Fix concatenate bug and improve performance

### PR DESCRIPTION
**Pull request by okuta**
This bug is from #452.
`concatenate` crash when it gets mixed dtype array.

And, this PR improve performance. The test code is [here](https://gist.github.com/okuta/ea732a251feea32ab80226ab6fa5ffe3)
```
# This PR #######################################################
shape: (256, 256, 256), axis: 0, order: C, elapsed: 16.1625622559
shape: (256, 256, 256), axis: 0, order: F, elapsed: 131.348251953
shape: (256, 256, 256), axis: 1, order: C, elapsed: 16.1895678711
shape: (256, 256, 256), axis: 1, order: F, elapsed: 130.555195312
shape: (256, 256, 256), axis: 2, order: C, elapsed: 16.4942749023
shape: (256, 256, 256), axis: 2, order: F, elapsed: 131.340351562

shape: (128, 128, 128), axis: 0, order: C, elapsed: 2.06588348389
shape: (128, 128, 128), axis: 0, order: F, elapsed: 10.6780529785
shape: (128, 128, 128), axis: 1, order: C, elapsed: 2.08701797485
shape: (128, 128, 128), axis: 1, order: F, elapsed: 9.83940124512
shape: (128, 128, 128), axis: 2, order: C, elapsed: 2.1305078125
shape: (128, 128, 128), axis: 2, order: F, elapsed: 9.80235351563

shape: (64, 64, 64), axis: 0, order: C, elapsed: 0.328061103821
shape: (64, 64, 64), axis: 0, order: F, elapsed: 0.5101404953
shape: (64, 64, 64), axis: 1, order: C, elapsed: 0.356145591736
shape: (64, 64, 64), axis: 1, order: F, elapsed: 0.518843193054
shape: (64, 64, 64), axis: 2, order: C, elapsed: 0.355208320618
shape: (64, 64, 64), axis: 2, order: F, elapsed: 0.549785270691

shape: (32, 32, 32), axis: 0, order: C, elapsed: 0.138884162903
shape: (32, 32, 32), axis: 0, order: F, elapsed: 0.371274871826
shape: (32, 32, 32), axis: 1, order: C, elapsed: 0.114077119827
shape: (32, 32, 32), axis: 1, order: F, elapsed: 0.483764801025
shape: (32, 32, 32), axis: 2, order: C, elapsed: 0.11232000351
shape: (32, 32, 32), axis: 2, order: F, elapsed: 0.501377601624


# master branch #################################################
shape: (256, 256, 256), axis: 0, order: C, elapsed: 16.3996728516
shape: (256, 256, 256), axis: 0, order: F, elapsed: 218.83328125
shape: (256, 256, 256), axis: 1, order: C, elapsed: 16.1702990723
shape: (256, 256, 256), axis: 1, order: F, elapsed: 130.665185547
shape: (256, 256, 256), axis: 2, order: C, elapsed: 79.9679003906
shape: (256, 256, 256), axis: 2, order: F, elapsed: 131.264052734

shape: (128, 128, 128), axis: 0, order: C, elapsed: 2.10899291992
shape: (128, 128, 128), axis: 0, order: F, elapsed: 24.5056542969
shape: (128, 128, 128), axis: 1, order: C, elapsed: 2.08578781128
shape: (128, 128, 128), axis: 1, order: F, elapsed: 9.82618713379
shape: (128, 128, 128), axis: 2, order: C, elapsed: 9.98044372559
shape: (128, 128, 128), axis: 2, order: F, elapsed: 9.78024780273

shape: (64, 64, 64), axis: 0, order: C, elapsed: 0.316738872528
shape: (64, 64, 64), axis: 0, order: F, elapsed: 1.84249404907
shape: (64, 64, 64), axis: 1, order: C, elapsed: 0.397792015076
shape: (64, 64, 64), axis: 1, order: F, elapsed: 0.517691841125
shape: (64, 64, 64), axis: 2, order: C, elapsed: 1.40936187744
shape: (64, 64, 64), axis: 2, order: F, elapsed: 0.560832328796

shape: (32, 32, 32), axis: 0, order: C, elapsed: 0.23242816925
shape: (32, 32, 32), axis: 0, order: F, elapsed: 0.361150093079
shape: (32, 32, 32), axis: 1, order: C, elapsed: 0.412528953552
shape: (32, 32, 32), axis: 1, order: F, elapsed: 0.380282897949
shape: (32, 32, 32), axis: 2, order: C, elapsed: 0.341022720337
shape: (32, 32, 32), axis: 2, order: F, elapsed: 0.382827835083

# v2 branch ######################################################
shape: (256, 256, 256), axis: 0, order: C, elapsed: 82.4052929688
shape: (256, 256, 256), axis: 0, order: F, elapsed: 261.209726563
shape: (256, 256, 256), axis: 1, order: C, elapsed: 82.3436035156
shape: (256, 256, 256), axis: 1, order: F, elapsed: 167.829375
shape: (256, 256, 256), axis: 2, order: C, elapsed: 83.3973535156
shape: (256, 256, 256), axis: 2, order: F, elapsed: 302.420722656

shape: (128, 128, 128), axis: 0, order: C, elapsed: 15.1660290527
shape: (128, 128, 128), axis: 0, order: F, elapsed: 22.8101855469
shape: (128, 128, 128), axis: 1, order: C, elapsed: 14.8621643066
shape: (128, 128, 128), axis: 1, order: F, elapsed: 22.6841137695
shape: (128, 128, 128), axis: 2, order: C, elapsed: 14.8460644531
shape: (128, 128, 128), axis: 2, order: F, elapsed: 23.0372827148

shape: (64, 64, 64), axis: 0, order: C, elapsed: 1.42650680542
shape: (64, 64, 64), axis: 0, order: F, elapsed: 1.70116027832
shape: (64, 64, 64), axis: 1, order: C, elapsed: 1.42875076294
shape: (64, 64, 64), axis: 1, order: F, elapsed: 3.27464080811
shape: (64, 64, 64), axis: 2, order: C, elapsed: 1.42323074341
shape: (64, 64, 64), axis: 2, order: F, elapsed: 3.39681365967

shape: (32, 32, 32), axis: 0, order: C, elapsed: 1.17972061157
shape: (32, 32, 32), axis: 0, order: F, elapsed: 0.228332805634
shape: (32, 32, 32), axis: 1, order: C, elapsed: 1.17656227112
shape: (32, 32, 32), axis: 1, order: F, elapsed: 0.34329536438
shape: (32, 32, 32), axis: 2, order: C, elapsed: 1.18050048828
shape: (32, 32, 32), axis: 2, order: F, elapsed: 0.355680656433
```

### CODE REVIEW

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -2440,7 +2440,8 @@ cpdef ndarray concatenate_method(tup, int axis):
     ndim = -1
     dtype = None
     have_same_types = True
-    for o in tup:
+    arrays = list(tup)
+    for o in arrays:
```
Comment (unnonouno) [inline cupy/core/core.pyx]: Is this because that `list` is faster than `tuple`?

File: cupy/core/core.pyx [hunk 2]
```diff
@@ -2472,100 +2473,115 @@ cpdef ndarray concatenate_method(tup, int axis):
...
 
-cpdef ndarray concatenate(tup, axis, shape, dtype):
+cpdef ndarray _concatenate(list arrays, Py_ssize_t axis, tuple shape, dtype):
+    cdef ndarray a, ret
+    cdef int i
+    cdef bint all_same_type, same_shape_and_contiguous
+    cdef Py_ssize_t axis_size
+    cdef Py_ssize_t threshold_size = 2 * 1024 * 1024
+
+    if len(arrays) > 8:
```
Comment (unnonouno) [inline cupy/core/core.pyx]: Why is this condition required? When an array is small, is this procedure inefficient?

File: cupy/core/core.pyx [hunk 3]
```diff
@@ -2472,100 +2473,115 @@ cpdef ndarray concatenate_method(tup, int axis):
...
+    return _concatenate(arrays, axis, tuple(shape), dtype)
 
 
-cpdef ndarray concatenate(tup, axis, shape, dtype):
+cpdef ndarray _concatenate(list arrays, Py_ssize_t axis, tuple shape, dtype):
+    cdef ndarray a, ret
+    cdef int i
+    cdef bint all_same_type, same_shape_and_contiguous
+    cdef Py_ssize_t axis_size
+    cdef Py_ssize_t threshold_size = 2 * 1024 * 1024
```
Comment (unnonouno) [inline cupy/core/core.pyx]: What does it mean? Please write a comment.

File: cupy/core/core.pyx (hunk 1 above)
Comment (okuta) [reply cupy/core/core.pyx]: Because `_concatinate` receives list.

File: cupy/core/core.pyx (hunk 2 above)
Comment (okuta) [reply cupy/core/core.pyx]: Yes.  this procedure needs host to device copy that takes a few time.

File: cupy/core/core.pyx (hunk 1 above)
Comment (unnonouno) [reply cupy/core/core.pyx]: i see

**Comment by unnonouno** [comment]
LGTM

**Comment by unnonouno** [comment]
Do we need to backport this fix? v2 did not merge #452

**Comment by okuta** [comment]
OK, I understand.

**Pull request merged**